deployment/environment/devenv/chain.go

```diff
@@ -94,58 +96,64 @@ func (c *ChainConfig) SetDeployerKey(pvtKeyStr *string) error {
 
 func NewChains(logger logger.Logger, configs []ChainConfig) (map[uint64]deployment.Chain, error) {
 	chains := make(map[uint64]deployment.Chain)
+	g := new(errgroup.Group)
 	for _, chainCfg := range configs {
-		selector, err := chainselectors.SelectorFromChainId(chainCfg.ChainID)
-		if err != nil {
-			return nil, fmt.Errorf("failed to get selector from chain id %d: %w", chainCfg.ChainID, err)
-		}
-		// TODO : better client handling
-		var ec *ethclient.Client
-		for _, rpc := range chainCfg.WSRPCs {
-			ec, err = ethclient.Dial(rpc.External)
+		chainCfg := chainCfg
+		g.Go(func() error {
+			selector, err := chainselectors.SelectorFromChainId(chainCfg.ChainID)
 			if err != nil {
-				logger.Warnf("failed to dial ws rpc %s", rpc)
-				continue
+				return fmt.Errorf("failed to get selector from chain id %d: %w", chainCfg.ChainID, err)
 			}
-			logger.Infof("connected to ws rpc %s", rpc)
-			break
-		}
-		if ec == nil {
-			return nil, fmt.Errorf("failed to connect to chain %s", chainCfg.ChainName)
-		}
-		chainInfo, err := deployment.ChainInfo(selector)
-		if err != nil {
-			return nil, fmt.Errorf("failed to get chain info for chain %s: %w", chainCfg.ChainName, err)
-		}
-		chains[selector] = deployment.Chain{
-			Selector:    selector,
-			Client:      ec,
-			DeployerKey: chainCfg.DeployerKey,
-			Confirm: func(tx *types.Transaction) (uint64, error) {
-				var blockNumber uint64
-				if tx == nil {
-					return 0, fmt.Errorf("tx was nil, nothing to confirm chain %s", chainInfo.ChainName)
-				}
-				ctx, cancel := context.WithTimeout(context.Background(), 3*time.Minute)
-				defer cancel()
-				receipt, err := bind.WaitMined(ctx, ec, tx)
+			// TODO : better client handling
+			var ec *ethclient.Client
+			for _, rpc := range chainCfg.WSRPCs {
+				ec, err = ethclient.Dial(rpc.External)
 				if err != nil {
-					return blockNumber, fmt.Errorf("failed to get confirmed receipt for chain %s: %w", chainInfo.ChainName, err)
+					logger.Warnf("failed to dial ws rpc %s: %s", rpc, err)
+					continue
 				}
-				if receipt == nil {
-					return blockNumber, fmt.Errorf("receipt was nil for tx %s chain %s", tx.Hash().Hex(), chainInfo.ChainName)
-				}
-				blockNumber = receipt.BlockNumber.Uint64()
-				if receipt.Status == 0 {
-					errReason, err := deployment.GetErrorReasonFromTx(ec, chainCfg.DeployerKey.From, tx, receipt)
-					if err == nil && errReason != "" {
-						return blockNumber, fmt.Errorf("tx %s reverted,error reason: %s chain %s", tx.Hash().Hex(), errReason, chainInfo.ChainName)
+				logger.Infof("connected to ws rpc %s", rpc)
+				break
+			}
+			if ec == nil {
+				return fmt.Errorf("failed to connect to chain %s", chainCfg.ChainName)
+			}
+			chainInfo, err := deployment.ChainInfo(selector)
+			if err != nil {
+				return fmt.Errorf("failed to get chain info for chain %s: %w", chainCfg.ChainName, err)
+			}
+			chains[selector] = deployment.Chain{
```

Comment:
this does not look like a thread safe operation? You will get concurrent write op error here.